tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -37,6 +39,40 @@ fn prefix_sum_carry_propagation(msb: u64, lsb: u64) -> u64 {
     }
 }
 
+fn should_hillis_steele_propagation_be_faster(num_blocks: usize, num_threads: usize) -> bool {
+    // Measures have shown that using a parallelized algorithm degrades
+    // the latency of a PBS, so we take that into account.
+    // (This factor is a bit pessimistic).
+    const PARALLEL_LATENCY_PENALTY: usize = 2;
```

Comment:
should we consider the worst case for parallelization in that case ? As we don't necesseraliy know if the penalty is the same ? As you point out it's a bit pessimisitc, realistically the only way to know for sure would be to benchmark but it's not something we'll do right away (there are some thoughts on that for a functionality on auto tuning down the line)

Reply:
What is the worst cae you think about?

Reply:
the worst case you are talking of here, it's an uncertain worst case is what I'm trying to say, even if we see empirically it's mostly true

Reply:
I think keeping that worst case penalty is ok as thats something we always observed for now

I'll rerun the tests without taking into account that penalty and see if it changes the chosen algorithms.

If it does change them then we will have to keep the penalty thing for sure

Reply:
Ok

Reply:
I reran the tests, witout the penalty factor the only change is that for 12 threads, 8 blocks it selects the parallel version (with the penalty factor it selects the non parallel one)

From benchmarks they are close (the parallel version being 5-7% slower)

Reply:
Actually ran the thing without penalty, it chooses the paralle version for all block size using 12 threads which does not match  benchmarks results

Reply:
so better with penalty ? my only concern is potentially having a bad surprise with large thread count, I don't know what the graph of those functions look like

Reply:
ah I think I know why, it thinks the whole layer is parallelized, while for small core counts it will get serialized by chunks of 12 e.g.

Reply:
so the true latency per layer is num_pbs_per_layer / thread_count (with some rounding, ceil or floor in there)

Reply:
> so better with penalty ? my only concern is potentially having a bad surprise with large thread count, I don't know what the graph of those functions look like

With the penalty we get a choice that matches performances measured by benchmarks

> so the true latency per layer is num_pbs_per_layer / thread_count (with some rounding, ceil or floor in there)
We do one PBS per block, so divide_ceil(num_blocks, num_threads) is a somewhat good estimation i would say

Reply:
ah yeah indeed, it's already in there, then let's go with the empirical formula for now and revisit if we see that these things change